app_build_suite/build_steps/helm.py

```diff
@@ -382,6 +403,72 @@ def pre_run(self, config: argparse.Namespace) -> None:
                     raise ValidationError(self.name, f"Value of '{option}' is not a correct boolean.")
 
+    def additional_annotations(self, chart_file_name: str, catalog_name: str) -> Dict[str, Any]:
+        catalog_base_url = f"{self._value_annotation_catalog_base_url}{catalog_name}/{chart_file_name}-meta/"
+        annotations = {}
+        # TODO check if these exist
+        annotations[self._key_annotation_metadata_url] = f"{catalog_base_url}main.yaml"
+        annotations[self._key_annotation_readme_url] = f"{catalog_base_url}README.md"
+        return annotations
+
+    def run(self, config: argparse.Namespace, context: Dict[str, Any]) -> None:
+        if not config.generate_metadata:
+            logger.info("Metadata generation is disabled using 'generate-metadata' option.")
+            return
+        # read current Chart.yaml
+        chart_yaml_path = os.path.join(config.chart_dir, _chart_yaml)
+        with open(chart_yaml_path, "r") as file:
+            chart_yaml = yaml.safe_load(file)
+        # try to guess the package file name. we need it for url generation in annotations
+        chart_name = os.path.basename(config.chart_dir)
+        context[context_key_chart_file_name] = f"{chart_name}-{context[context_key_git_version]}.tgz"
+        # put in generated annotations
+        chart_yaml[self._key_annotations] = {
+            **chart_yaml.get(self._key_annotations, {}),
+            **self.additional_annotations(context[context_key_chart_file_name], config.catalog_name),
+        }
+        # save Chart.yaml
+        save_yaml(chart_yaml_path, chart_yaml)
+
+
+class HelmChartMetadataFinalizer(BuildStep):
+    """
+    HelmChartMetadataFinalizer finalizes metadata generation based on additional info in Chart.yaml file
+    Should run after HelmChartBuilder
+    """
+
+    _key_upstream_chart_url = "upstreamChartURL"
+    _key_restrictions = "restrictions"
+    _key_cluster_singleton = "clusterSingleton"
+    _key_namespace_singleton = "namespaceSingleton"
+    _key_gpu_instances = "gpuInstances"
+    _key_fixed_namespace = "fixedNamespace"
+    _key_chart_file = "chartFile"
+    _key_digest = "digest"
+    _key_date_created = "dateCreated"
+    _key_chart_api_version = "chartApiVersion"
+    _key_api_version = "apiVersion"
+    _key_annotations = "annotations"
+
+    @property
+    def steps_provided(self) -> Set[StepType]:
+        return {STEP_METADATA}
+
+    def pre_run(self, config: argparse.Namespace) -> None:
+        if not config.generate_metadata:
+            logger.info("Metadata generation is disabled using 'generate-metadata' option.")
+            return
+        # first step of validation should be done already by 'ct' with correct schema (unless explicitly disabled)
+        chart_yaml_path = os.path.join(config.chart_dir, _chart_yaml)
+        with open(chart_yaml_path, "r") as file:
+            chart_yaml = yaml.safe_load(file)
+        # TODO decide if we want to validate the new annotations here
```

Comment:
I don't think we have to, since we're generating them automatically and overwriting anyway.